### PR TITLE
feat: add `lightclient` module and migrate block header state in this module

### PR DESCRIPTION
**Pull request by lumtis**
# Description

- Initialize `lightclient` module to offload the block header data currently in `observer` there
- Rename `BlockHeaderState` into `ChainState` to avoid confusion with "having the state of a block header" since this represents the overall state for block headers for a specific chains
- Move the verification flags values from `crosschainFlags` to `verificationFlags` part of the `lightclient` module state
- Rename `AddBlockHeader` in `VoteBlockHeader` for `observer`
- Move verification and block adding logic from VoteBlockHeader to the `lightclient` module
- Message to update `VerificationFlags`
- Unit tests
- add breaking changes notice

NOTE: I didn't include a migration script. Currently the block headers features is disabled on live networks. Introducing this module will simply set the state at 0. Including a migration script would introduce risks without benefits as there is no block headers for mainnet

Closes: https://github.com/zeta-chain/node/issues/1491 https://github.com/zeta-chain/node/issues/1975

### CODE REVIEW

File: e2e/runner/setup_zeta.go [hunk 1]
```diff
@@ -204,6 +204,13 @@ func (runner *E2ERunner) SetupBTCZRC20() {
 	runner.BTCZRC20 = BTCZRC20
 }
 
+// EnableVerificationFlags enables the verification flags on ZetaChain
+func (runner *E2ERunner) EnableVerificationFlags() error {
+	runner.Logger.Print("⚙️ enabling verification flags for block headers")
+
+	return runner.ZetaTxServer.EnableVerificationFlags(e2eutils.FungibleAdminName)
```
[the remark this answers is not in the document]
Comment (lumtis) [reply e2e/runner/setup_zeta.go:211]: Yeah I think we used to only call admin func in fungible module. Let's rename it

File: docs/spec/lightclient/messages.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Messages
+
+## MsgUpdateVerificationFlags
+
+UpdateVerificationFlags updates the crosschain related flags.
```
[the remark this answers is not in the document]
Comment (lumtis) [reply docs/spec/lightclient/messages.md:5]: No, c/c mistake

File: proto/observer/crosschain_flags.proto [hunk 3]
```diff
@@ -21,6 +21,8 @@ message GasPriceIncreaseFlags {
   // Maximum number of pending crosschain transactions to check for gas price increase
   uint32 maxPendingCctxs = 5;
 }
+
+// Deprecated(v16): Use VerificationFlags in the lightclient module instead
```
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/observer/crosschain_flags.proto:25]: Yeah, it is disabled on both testnet and mainnet so it would be a noop

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker.go [hunk 4]
```diff
@@ -36,180 +34,110 @@ func (k msgServer) AddToOutTxTracker(goCtx context.Context, msg *types.MsgAddToO
...
+	// add the tracker to the list
+	tracker.HashList = append(tracker.HashList, &hash)
+	k.SetOutTxTracker(ctx, tracker)
 	return &types.MsgAddToOutTxTrackerResponse{}, nil
 }
 
-func (k Keeper) VerifyOutTxBody(ctx sdk.Context, msg *types.MsgAddToOutTxTracker, txBytes []byte) error {
+// verifyProofAndOutTxBody verifies the proof and outbound tx body
+func verifyProofAndOutTxBody(ctx sdk.Context, k msgServer, msg *types.MsgAddToOutTxTracker) error {
+	txBytes, err := k.lightclientKeeper.VerifyProof(ctx, msg.Proof, msg.ChainId, msg.BlockHash, msg.TxIndex)
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/keeper/msg_server_add_to_outtx_tracker.go:116]: It's an internal function only used in the function above where it is checked so I would consider this as a precondition to be fine but we can precise this precondition in comment

File: e2e/runner/setup_zeta.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply e2e/runner/setup_zeta.go]: bd0ab6f5

File: docs/spec/lightclient/messages.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply docs/spec/lightclient/messages.md]: bd0ab6f5

File: x/lightclient/client/cli/query.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/spf13/cobra"
+	"github.com/zeta-chain/zetacore/x/lightclient/types"
+)
+
+// GetQueryCmd returns the cli query commands for this module
+func GetQueryCmd(_ string) *cobra.Command {
+	// Group crosschain queries under a subcommand
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/lightclient/client/cli/query.go]: bd0ab6f5

File: x/lightclient/types/message_update_crosschain_flags.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+const (
+	TypeMsgUpdateVerificationFlags = "update_verification_flags"
+)
+
+var _ sdk.Msg = &MsgUpdateVerificationFlags{}
+
+func NewMsgUpdateVerificationFlags(creator string, ethTypeChainEnabled, btcTypeChainEnabled bool) *MsgUpdateVerificationFlags {
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/lightclient/types/message_update_crosschain_flags.go]: bd0ab6f5

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/keeper/msg_server_add_to_outtx_tracker.go:119]: bd0ab6f5

File: zetaclient/bitcoin/bitcoin_client.go [hunk 7]
```diff
@@ -436,8 +436,8 @@ func (ob *BTCChainClient) ObserveInTx() error {
 		// add block header to zetabridge
 		// TODO: consider having a separate ticker(from TSS scaning) for posting block headers
 		// https://github.com/zeta-chain/node/issues/1847
-		flags := ob.coreContext.GetCrossChainFlags()
-		if flags.BlockHeaderVerificationFlags != nil && flags.BlockHeaderVerificationFlags.IsBtcTypeChainEnabled {
+		verificationFlags := ob.coreContext.GetVerificationFlags()
```
[the remark this answers is not in the document]
Comment (lumtis) [reply zetaclient/bitcoin/bitcoin_client.go:439]: Yeah I didn't really have idea on this one. Open for suggestions

**Comment by lumtis** [comment]
>just reminder to check nosec lint issue before merge

This is a bug from the action itself

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker.go [hunk 8]
```diff
@@ -36,180 +34,113 @@ func (k msgServer) AddToOutTxTracker(goCtx context.Context, msg *types.MsgAddToO
...
-		bitcoinNetParams, err := chains.BitcoinNetParamsFromChainID(msg.ChainId)
-		if err != nil {
-			return fmt.Errorf("failed to get Bitcoin net params, error %s", err.Error())
-		}
-		addrP2WPKH, err := btcutil.NewAddressWitnessPubKeyHash(
-			btcutil.Hash160(pubKey.SerializeCompressed()),
-			bitcoinNetParams,
-		)
+	if err != nil || tss == nil {
+		reason := "tss response is nil"
```
Comment (kingpinXD) [inline x/crosschain/keeper/msg_server_add_to_outtx_tracker.go:134]: I feel like this can be simplified into just returning an error if there is an error and then checking if TSS is nil.

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker_test.go [hunk 9]
```diff
@@ -17,89 +22,173 @@ func getEthereumChainID() int64 {
...
-	cctx := types.CrossChainTx{
-		Creator: "any",
-		Index:   "0x123",
-		CctxStatus: &types.Status{
-			Status: status,
-		},
+func TestMsgServer_AddToOutTxTracker(t *testing.T) {
+
+	// mockCctxByNonce mocks the methods called by CctxByNonce to directly return the given cctx or error
+	mockCctxByNonce := func(
```
Comment (kingpinXD) [inline x/crosschain/keeper/msg_server_add_to_outtx_tracker_test.go:31]: Would be a good idea to move this function into the test-utils package

File: x/crosschain/types/tx_body_verification.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		return fmt.Errorf("invalid nonce, want nonce %d, got %d", txx.Nonce(), msg.Nonce)
+	}
+	if txx.Hash().Hex() != msg.TxHash {
+		return fmt.Errorf("invalid tx hash, want tx hash %s, got %s", txx.Hash().Hex(), msg.TxHash)
+	}
+	return nil
+}
+
+// verifyOutTxBodyBTC validates the SegWit sender address, nonce and chain id and tx hash
+func verifyOutTxBodyBTC(msg MsgAddToOutTxTracker, txBytes []byte, tssBtc string) error {
```
Comment (kingpinXD) [inline x/crosschain/types/tx_body_verification.go:121]: Not required for this pr but we should add tests to improve the code coverage sections pointed out by codecov, we can can add a issue to track it

File: x/lightclient/genesis.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// ExportGenesis returns the lightclient module's exported genesis.
+func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
+	verificationFlags, found := k.GetVerificationFlags(ctx)
+	if !found {
+		verificationFlags = types.VerificationFlags{
+			EthTypeChainEnabled: false,
+			BtcTypeChainEnabled: false,
+		}
```
Comment (kingpinXD) [inline x/lightclient/genesis.go:32]: We can create a fucntion `DefaultVerficationFlags` and use it here and in `DefaultGenesisState`

File: x/lightclient/keeper/chain_state.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+// GetChainState returns a chain state from its chainID
+func (k Keeper) GetChainState(ctx sdk.Context, chainID int64) (val types.ChainState, found bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ChainStateKey))
+
+	b := store.Get(types.KeyPrefix(strconv.FormatInt(chainID, 10)))
+	if b == nil {
+		return val, false
+	}
+
+	k.cdc.MustUnmarshal(b, &val)
```
Comment (kingpinXD) [inline x/lightclient/keeper/chain_state.go:47]: Lets change this to 
`Unmarshal` and handle the error to be safe

File: x/lightclient/keeper/chain_state.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+		list = append(list, val)
+	}
+
+	return list
+}
+
+// SetChainState set a specific chain state in the store from its index
+func (k Keeper) SetChainState(ctx sdk.Context, chainState types.ChainState) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ChainStateKey))
+	b := k.cdc.MustMarshal(&chainState)
```
Comment (kingpinXD) [inline x/lightclient/keeper/chain_state.go:33]: Lets also change this to `Marshal`

File: x/lightclient/keeper/grpc_query_block_header.go [hunk 14]
```diff
@@ -1,35 +1,33 @@
...
+func (k Keeper) BlockHeaderAll(c context.Context, req *types.QueryAllBlockHeaderRequest) (*types.QueryAllBlockHeaderResponse, error) {
 	if req == nil {
 		return nil, status.Error(codes.InvalidArgument, "invalid request")
 	}
 	ctx := sdk.UnwrapSDKContext(c)
 	store := ctx.KVStore(k.storeKey)
 	blockHeaderStore := prefix.NewStore(store, types.KeyPrefix(types.BlockHeaderKey))
 
-	var blockHeaders []*proofs.BlockHeader
+	var blockHeaders []proofs.BlockHeader
```
Comment (kingpinXD) [inline x/lightclient/keeper/grpc_query_block_header.go:23]: NIT : we can add a function GetBlockHeadersPaginated , it would be easier to unit test

File: x/lightclient/keeper/grpc_query_chain_state.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+func (k Keeper) ChainStateAll(c context.Context, req *types.QueryAllChainStateRequest) (*types.QueryAllChainStateResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+	ctx := sdk.UnwrapSDKContext(c)
+	store := ctx.KVStore(k.storeKey)
+	chainStateStore := prefix.NewStore(store, types.KeyPrefix(types.ChainStateKey))
+
+	var chainStates []types.ChainState
+	pageRes, err := query.Paginate(chainStateStore, req.Pagination, func(key []byte, value []byte) error {
```
Comment (kingpinXD) [inline x/lightclient/keeper/grpc_query_chain_state.go:25]: NIT : same as block headers , we can add a function `GetChainStatePaginated`

File: x/lightclient/keeper/msg_server_update_verification_flags.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authoritytypes "github.com/zeta-chain/zetacore/x/authority/types"
+	"github.com/zeta-chain/zetacore/x/lightclient/types"
+)
+
+// UpdateVerificationFlags updates the light client verification flags.
+// This disables/enables blocks verification of the light client for the specified chain.
+// Emergency group can disable flags while operation group can enable/disable
```
Comment (kingpinXD) [inline x/lightclient/keeper/msg_server_update_verification_flags.go:13]: Is this correct? Maybe we should update this to be more accurate 
Also, mention the condition where one is being enabled and the other is being disabled ( which would still require the Operational group)

File: x/lightclient/keeper/msg_server_update_verification_flags.go [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+func (k msgServer) UpdateVerificationFlags(goCtx context.Context, msg *types.MsgUpdateVerificationFlags) (
+	*types.MsgUpdateVerificationFlagsResponse,
+	error,
+) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	requiredGroup := authoritytypes.PolicyType_groupEmergency
+	if msg.VerificationFlags.EthTypeChainEnabled || msg.VerificationFlags.BtcTypeChainEnabled {
+		requiredGroup = authoritytypes.PolicyType_groupOperational
+	}
```
Comment (kingpinXD) [inline x/lightclient/keeper/msg_server_update_verification_flags.go:23]: NIT: We can move this to a separate function and add unit tests to it

File: x/lightclient/keeper/proof_test.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/zeta-chain/zetacore/pkg/proofs/ethereum"
+	"github.com/zeta-chain/zetacore/pkg/testdata"
+	keepertest "github.com/zeta-chain/zetacore/testutil/keeper"
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	"github.com/zeta-chain/zetacore/x/lightclient/types"
+)
+
+// generateProof generates a proof and block header
+// returns the proof, block header, block hash, tx index, chain id, and tx hash
+func generateProof(t *testing.T) (*proofs.Proof, proofs.BlockHeader, string, int64, int64, ethcommon.Hash) {
```
Comment (kingpinXD) [inline x/lightclient/keeper/proof_test.go:22]: This might also be a good candidate for a sample; I feel any test data-related functions should be moved to the sample directory.

File: x/observer/types/errors.go [hunk 19]
```diff
@@ -37,4 +34,5 @@ var (
 	ErrInboundDisabled      = errorsmod.Register(ModuleName, 1132, "inbound tx processing is disabled")
 	ErrInvalidZetaCoinTypes = errorsmod.Register(ModuleName, 1133, "invalid zeta coin types")
 	ErrNotObserver          = errorsmod.Register(ModuleName, 1134, "sender is not an observer")
+	ErrInvalidBlockHeader   = errorsmod.Register(ModuleName, 1135, "invalid block header")
```
Comment (kingpinXD) [inline x/observer/types/errors.go:37]: imo , Errors related to a module should be defined in that module itself, even if other modules use them. 
Right now, only the observer module is using this, but in the future, we might use this in other places 

 we should be doing this for other modules as well , might be a good idea to create a task to trackit

File: cmd/zetae2e/local/accounts.go [hunk 20]
```diff
@@ -34,5 +34,5 @@ var (
 	UserAdminAddress    = ethcommon.HexToAddress("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266")
 	UserAdminPrivateKey = "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80" // #nosec G101 - used for testing
 
-	FungibleAdminMnemonic = "snow grace federal cupboard arrive fancy gym lady uniform rotate exercise either leave alien grass" // #nosec G101 - used for testing
+	AdminMnemonic = "snow grace federal cupboard arrive fancy gym lady uniform rotate exercise either leave alien grass" // #nosec G101 - used for testing
```
Comment (kingpinXD) [inline cmd/zetae2e/local/accounts.go:37]: I think we should keep the name `FungibleAdmin`. The same identifier is used in other reports, such as protocol contracts as well . Having the same name makes reading and debugging code easier when working across repos.

With that said however , I do agree with using just Admin , but maybe we should do it in a separate PR , nad rename the functions on the smart contracts as well

File: e2e/utils/zetacore.go [hunk 21]
```diff
@@ -11,7 +11,7 @@ import (
 )
 
 const (
-	FungibleAdminName = "fungibleadmin"
+	AdminName = "fadmin"
```
Comment (kingpinXD) [inline e2e/utils/zetacore.go:14]: Same as my previous comment , we should keep this as it is

File: cmd/zetae2e/local/accounts.go (hunk 20 above)
Comment (lumtis) [reply cmd/zetae2e/local/accounts.go:37]: Ok

Comment (lumtis) [reply cmd/zetae2e/local/accounts.go]: 58b2d923

File: e2e/utils/zetacore.go (hunk 21 above)
Comment (lumtis) [reply e2e/utils/zetacore.go]: 58b2d923

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker_test.go (hunk 9 above)
Comment (lumtis) [reply x/crosschain/keeper/msg_server_add_to_outtx_tracker_test.go]: a5f9653a

File: x/crosschain/types/tx_body_verification.go (hunk 10 above)
Comment (lumtis) [reply x/crosschain/types/tx_body_verification.go:121]: Yeah, created this issue before: https://github.com/zeta-chain/node/issues/1994
I added a todo to track it in this file to makes it clearer

File: x/lightclient/keeper/chain_state.go (hunk 12 above)
Comment (lumtis) [reply x/lightclient/keeper/chain_state.go:47]: This is a good point.

We have this for every state variables now, so I created an issue for this: https://github.com/zeta-chain/node/issues/2035

I think we can address it for this variable in this issue. I don't see many risk as the functionality would be disabled for now, so we can address it once we plan to enable block header verification

File: x/lightclient/keeper/chain_state.go (hunk 13 above)
Comment (lumtis) [reply x/lightclient/keeper/chain_state.go:33]: Same comment https://github.com/zeta-chain/node/issues/2035

File: x/lightclient/keeper/grpc_query_block_header.go (hunk 14 above)
Comment (lumtis) [reply x/lightclient/keeper/grpc_query_block_header.go:23]: I wouldn't see it as necessary since the query is the only function using the pagination option and the function is just doing this so the query would just call this function but I think we can add better test for pagination for the query

File: x/crosschain/types/tx_body_verification.go (hunk 10 above)
Comment (lumtis) [reply x/crosschain/types/tx_body_verification.go:123]: b5578adc

File: x/lightclient/genesis.go (hunk 11 above)
Comment (lumtis) [reply x/lightclient/genesis.go]: 46eabcc5

File: x/lightclient/keeper/grpc_query_block_header.go (hunk 14 above)
Comment (lumtis) [reply x/lightclient/keeper/grpc_query_block_header.go:23]: For the tests: 1ee85715

File: x/lightclient/keeper/grpc_query_chain_state.go (hunk 15 above)
Comment (lumtis) [reply x/lightclient/keeper/grpc_query_chain_state.go:25]: 1ee85715

File: x/lightclient/keeper/msg_server_update_verification_flags.go (hunk 16 above)
Comment (lumtis) [reply x/lightclient/keeper/msg_server_update_verification_flags.go]: Updated the comment 785e788c

File: x/lightclient/keeper/msg_server_update_verification_flags.go (hunk 17 above)
Comment (lumtis) [reply x/lightclient/keeper/msg_server_update_verification_flags.go]: c2c0409c

File: x/lightclient/keeper/proof_test.go (hunk 18 above)
Comment (lumtis) [reply x/lightclient/keeper/proof_test.go]: Sounds good 60e57f36

File: x/observer/types/errors.go (hunk 19 above)
Comment (lumtis) [reply x/observer/types/errors.go]: 5c5bc0d5

File: x/crosschain/keeper/msg_server_add_to_outtx_tracker.go (hunk 8 above)
Comment (lumtis) [reply x/crosschain/keeper/msg_server_add_to_outtx_tracker.go]: You mean removing the ErrTssNotFound wrapper? Why?
